packages/core-data/src/footnotes/get-footnotes-order.js

```diff
@@ -7,24 +12,31 @@ const cache = new WeakMap();
 
 function getBlockFootnotesOrder( block ) {
 	if ( ! cache.has( block ) ) {
-		const content = getRichTextValuesCached( block ).join( '' );
-		const newOrder = [];
-
-		// https://github.com/WordPress/gutenberg/pull/43204 lands. We can then
-		// get the order directly from the rich text values.
-		if ( content.indexOf( 'data-fn' ) !== -1 ) {
-			const regex = /data-fn="([^"]+)"/g;
-			let match;
-			while ( ( match = regex.exec( content ) ) !== null ) {
-				newOrder.push( match[ 1 ] );
-			}
-		}
-		cache.set( block, newOrder );
+		const order = getRichTextValuesCached( block ).reduce(
+			( acc, value ) => {
+				if ( value.indexOf( 'data-fn' ) !== -1 ) {
+					create( { html: value } ).replacements.forEach(
+						( { type, attributes } ) => {
+							if ( type === 'core/footnote' ) {
+								acc.push( attributes[ 'data-fn' ] );
+							}
+						}
+					);
+				}
+				return acc;
+			},
+			[]
+		);
```

Comment:
with as much setup as we have here, mixed with the imperative nature of the reduction, would it be simpler/clearer expressed as a `for` loop?

```js
const order = [];
for ( const value of getRichTextValuesCached( block ) ) {
	if ( ! value.includes( 'data-fn' ) ) {
		continue;
	}

	for ( const { type, attributes } of create( { html: value } ).replacements ) {
		if ( type === 'core/footnote' ) {
			order.push( attributes[ 'data-fn' ] );
		}
	}
}
```

Reply:
Had to revert this. For ... of doesn't skip the empty slots in the sparse array, so forEach works better here?

Reply:
> Had to revert this. For ... of doesn't skip the empty slots in the sparse array, so forEach works better here?

`.forEach` is fine, but just FYI `for ... in` skips the empty slots, if that's useful.

Reply:
Yep, but that just gives the index instead of the value